Hardware keyboard: Web, embedder, and dart:ui

dkwingsmt opened the pull request:
## Description

This PR is part of [Hardware Keyboard project](https://github.com/flutter/engine/pull/23465), specifically the embedder API, dart:ui API, and Web engine parts. The embedder and dart:ui parts are the foundation to other engine changes, while Web is required to change simultaneously for the API mirroring check.

The hardware keyboard API requires each platform to convert native events to "key data" that meets Flutter's specification and send them via the embedder API. The framework is then able to receive `ui.KeyData`s in a regular, unified form, and perform the simplest 1-to-1 conversion into `KeyEvent` (to be added) and assertions.

The key event model is described at [here](https://github.com/flutter/flutter/blob/79a4c94a7d8e891f3ce706699838a6c8c2adf39b/packages/flutter/lib/src/services/hardware_keyboard.dart#L293). Simply put,

- Key events are categorized into down, up, and repeat
- A key press always starts with a down, goes through zero or more repeats, then ends with an up. 
- All events throughout the press contains the same physical key and logical key.
- Missing native down or up events will be synthesized to guarantee the regularity.

### Public API changes

This PR adds the following APIs to dart:ui:
- enum `KeyChange`
- class `KeyData`
- typedef `KeyDataCallback`
- property `PlatformDispatcher.onKeyData`
- property `SingletonFlutterWindow.onKeyData`

This PR adds the following APIs to embedder.h:
- enum `FlutterKeyEventKind`
- struct `FlutterKeyEvent`
- function `FlutterEngineSendKeyEvent`
- typedef `FlutterEngineSendKeyEventFnPtr `
- property `FlutterEngineProcTable.SendKeyEvent`

Although the key mapping file claims to be auto-generated, the codegen script is not checked in until https://github.com/flutter/flutter/pull/73440.

## Related Issues

- Umbrella: flutter/flutter#44918

## Tests

I added the following tests:

For web engine:
- Single key press, repeat, and release
- Release modifier during a repeated sequence
- Distinguish between left and right modifiers
- Distinguish between normal and numpad digits
- Prevents default when pressing Tab
- Dead keys are distinguishable
- Duplicate down is ignored
- Duplicate ups are skipped
- Conflict from multiple keyboards do not crash
- CapsLock down synthesizes an immediate cancel on macOS
- CapsLock behaves normally on non-macOS
- Key guards: key down events are guarded
- Key guards: key repeated down events refreshes guards
- Key guards: cleared by keyups
- Lock flags of other keys
- Deduce modifier key cancel from modifier field

For embedder:
-  KeyDataIsCorrectlySerialized
- KeyDataResponseIsCorrectlyInvoked

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [contributor guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [C++, Objective-C, Java style guides] for the engine.
- [x] I read the [tree hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation.
- [x] All existing and new tests are passing.
- [x] I am willing to follow-up on review comments in a timely manner.


## Reviewer Checklist

- [ ] I have submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.


## Breaking Change

Did any tests fail when you ran them? Please read [handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: https://flutter.dev/go/breaking-changes-template *Replace this with a link to a pull request that adds the migration guide to https://flutter.dev/docs/release/breaking-changes*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[contributor guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[tree hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

dkwingsmt commented:
cc @gspencergoog

stuartmorgan commented on lib/ui/key.dart:
```diff
@@ -0,0 +1,85 @@
+
+  /// Time of event dispatch, relative to an arbitrary timeline.
+  ///
+  /// For [KeyChange.synchronize] and [KeyChange.cancel] events, the [timeStamp]
```
Weren't `KeyChange.synchronized` and `KeyChange.cancel` replaced by the `synthesized` property on this class?

stuartmorgan commented on lib/ui/key.dart:
```diff
@@ -0,0 +1,85 @@
+  final String? character;
+
+  /// Whether the event is generated by Flutter, hence not corresponding to a
+  /// native event.
```
I'm not sure this is a sufficient explanation for someone not versed in the design; I would expand this to clearly explain the kinds of cases where a synthesized event is sent.

stuartmorgan commented on lib/ui/window/key_data.h:
```diff
@@ -0,0 +1,48 @@
+
+// The fixed-length sections of a KeyDataPacket.
+//
+// KeyData does not contain `character`, for various-length data are stored in a
```
Nit: s/various/variable/

stuartmorgan commented on shell/platform/embedder/embedder.cc:
```diff
@@ -1492,6 +1492,66 @@ FlutterEngineResult FlutterEngineSendPointerEvent(
+      SAFE_ACCESS(event, kind, FlutterKeyEventKind::kFlutterKeyEventKindUp));
+  key_data.physical = SAFE_ACCESS(event, physical, 0);
+  key_data.logical = SAFE_ACCESS(event, logical, 0);
+  key_data.synthesized = !!SAFE_ACCESS(event, synthesized, false);
```
The only values for the C++ `bool` type are `true` and `false` (as opposed to, say old ObjC `BOOL` behavior), by spec, so you don't need to double-negate here. If you get any other value, a client has done something with undefined behavior.

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -611,6 +611,36 @@ typedef struct {
+typedef struct {
+  /// The size of this struct. Must be sizeof(FlutterKeyEvent).
+  size_t struct_size;
+  // Timestamp in microseconds.
```
The comment should be very clear about the requirements here. E.g., is using a monotonic clock a requirement, or is it okay for later events to have earlier timestamps when there's a time change?

dkwingsmt replied:
I see the time stamp of key events are defined as `FlutterEngineGetCurrentTime`. Currently I'm using as much time from native as possible in my PRs, wondering if it might provide extra precision or information to applications with certain requirement. Do you recommend me to switch to `FlutterEngineGetCurrentTime`?

stuartmorgan replied:
Not if we don't need it; I just think we should be explicit about what is (or isn't required). If we're okay with native event time, we should say that.

Is it the case that Flutter itself will be fine with non-monotonic event times?

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -611,6 +611,36 @@ typedef struct {
+  uint64_t physical;
+  // The logical key of the event, usually the effect of the key without
+  // modifier, but might include modifier if the information is not available.
+  // Can be 0 for empty.
```
It's not obvious what this means exactly; what is the definitive source of numbers for logical keys?

Since this API surface is intended for use by third-party developers as well, it's important not to assume too much context in the comments.

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -611,6 +611,36 @@ typedef struct {
+  // Can be 0 for empty.
+  uint64_t logical;
+  // Null-terminated character input from the event. Can be null. Not available
+  // to up events.
```
Nit: s/Not available to/ignored for/

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -611,6 +611,36 @@ typedef struct {
+  // Null-terminated character input from the event. Can be null. Not available
+  // to up events.
+  const char* character;
+  // Whether the event is synthesized by Flutter.
```
This comment is very confusing, since the embedder is providing this data, and it's not Flutter. It's whether it is synthesized by the embedder. And as with the above, this comment needs a lot more context: what is a synthesized event, and when should an embedder send one?

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -1549,6 +1579,12 @@ FlutterEngineResult FlutterEngineSendPointerEvent(
     size_t events_count);
 
+FLUTTER_EXPORT
+FlutterEngineResult FlutterEngineSendKeyEvent(
```
This should be documented.

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -611,6 +611,36 @@ typedef struct {
+
+struct _FlutterPlatformKeyEventResponseHandle;
+typedef struct _FlutterPlatformKeyEventResponseHandle
+    FlutterPlatformKeyEventResponseHandle;
```
There doesn't appear to be any explanation of what this is or what it's for.

stuartmorgan reviewed (changes requested):
I would expect to see tests for the new embedder APIs (i.e., changes in embedder/test/) as part of this.

mdebbar commented on lib/web_ui/lib/src/engine/keyboard_binding.dart:
```diff
@@ -0,0 +1,489 @@
+    if (_instance == null) {
+      _instance = KeyboardBinding._(glassPaneElement);
+      assert(() {
+        registerHotRestartListener(_instance!._reset);
```
`_reset()` calls `_setup()`. I don't think you want to setup twice during a hot restart, right?

The callback passed to `registerHotRestartListener` should only do clean up. Setup will be executed later when the `KeyboardBinding` instance is re-created.

mdebbar commented on lib/web_ui/lib/src/engine/keyboard_binding.dart:
```diff
@@ -0,0 +1,489 @@
+  bool handleEvent(FlutterHtmlKeyboardEvent event) {
+    final Duration timeStamp = _eventTimeStampToDuration(event.timeStamp!);
+
+    if (_shouldPreventDefault(event)) {
+      event.preventDefault();
+    }
```
We can get rid of this if we listen to the response from the framework. The response is a Map that has a `handled` key. If `handled` is `true`, we should call `event.preventDefault()`.

mdebbar replied:
Actually, @gspencergoog is doing this in https://github.com/flutter/engine/pull/23655.

mdebbar commented on lib/web_ui/lib/src/engine/keyboard_binding.dart:
```diff
@@ -0,0 +1,489 @@
+      change = ui.KeyChange.down;
+      if (lastLogicalRecord != null) {
+        // This physical key is being pressed according to the record.
+        if (event.repeat ?? false) {
```
Another pattern that I've seen is `if (event.repeat == true) { ... }`. I couldn't find any recommendation in the style guide, so I'm not sure which patter is preferred.

dkwingsmt replied:
Thanks for pointing out, although I do think this style is more readable.

mdebbar commented on lib/web_ui/lib/src/engine/keyboard_binding.dart:
```diff
@@ -0,0 +1,489 @@
+      change = ui.KeyChange.up;
+    }
+
+    late final int? nextLogicalRecord;
```
You can drop the `late` here, as long as the switch statement below is exhaustive and the static analyzer can verify that `nextLogicalRecord` is always initialized exactly one time.

mdebbar commented on lib/web_ui/lib/src/ui/key.dart:
```diff
@@ -0,0 +1,66 @@
+
+  /// Returns a complete textual description of the information in this object.
+  String toStringFull() {
+    return '$runtimeType('
+           ')';
```
Incomplete?

dkwingsmt commented:
@stuartmorgan I updated API, docs, and added embedder tests. Can you take a look and make a full review?

stuartmorgan commented on lib/ui/window/key_data_packet.h:
```diff
@@ -0,0 +1,47 @@
+  void SetCharacter(const char* characters);
+
+ private:
+  size_t CharacterSizeStart_() { return 0; }
```
These need comments. I'm also not clear why they are returning indexes rather than pointers given the way they are used.

stuartmorgan commented on lib/ui/window/key_data_packet.h:
```diff
@@ -0,0 +1,47 @@
+
+namespace flutter {
+
+typedef std::function<void (bool)> KeyDataPacketCallback;
```
This needs a comment. (I'm not even clear why it's here, since it's not used anywhere in this class.)

stuartmorgan commented on lib/ui/window/key_data_packet.h:
```diff
@@ -0,0 +1,47 @@
+
+  void SetKeyData(const KeyData& event);
+
+  // Set character data to the proper position, which should not be terminated
```
What does "to the proper position" mean? That sounds like it's keeping a reference to the pointer, but it's actually copying the data.

stuartmorgan commented on lib/ui/window/key_data_packet.h:
```diff
@@ -0,0 +1,47 @@
+
+  const std::vector<uint8_t>& data() const { return data_; }
+
+  void SetKeyData(const KeyData& event);
```
Needs a comment.

stuartmorgan commented on lib/ui/window/key_data_packet.h:
```diff
@@ -0,0 +1,47 @@
+
+  std::vector<uint8_t> data_;
+
+  FML_DISALLOW_COPY_AND_ASSIGN(KeyDataPacket);
```
This is no longer allowed by the style guide; the copy and assign constructors should be deleted in the public section.

stuartmorgan commented on lib/ui/window/key_data_packet.h:
```diff
@@ -0,0 +1,47 @@
+  void SetKeyData(const KeyData& event);
+
+  // Set character data to the proper position, which should not be terminated
+  // by a null character (length controled by character_data_size).
```
Having setting the length and setting the data being completely different operations seems dangerous. What's the purpose of this? The only use I see in the PR of constructing one of these objects all the data is set at the same time, so it seems like it would be much, much simpler from an API perspective to just take everything in the constructor. Is there another use case for this API?

stuartmorgan commented on lib/ui/window/platform_configuration.cc:
```diff
@@ -424,6 +442,20 @@ void PlatformConfiguration::CompletePlatformMessageResponse(
 }
 
+void PlatformConfiguration::CompleteKeyDataResponse(uint64_t response_id, bool handled) {
+  if (!response_id) {
```
Nit: If you want to compare an int to zero, use `== 0` rather than `!` so that it doesn't look like a bool or pointer.

stuartmorgan commented on lib/ui/window/platform_configuration.cc:
```diff
@@ -424,6 +442,20 @@ void PlatformConfiguration::CompletePlatformMessageResponse(
+  if (it == pending_key_responses_.end()) {
+    return;
+  }
+  auto callback = std::move(it->second);
```
The type for this `auto` isn't obvious locally, so per style guide prefer using an explicit type.

stuartmorgan commented on lib/ui/window/platform_configuration.h:
```diff
@@ -323,6 +323,24 @@ class PlatformConfiguration final {
                                std::vector<uint8_t> args);
 
+  //----------------------------------------------------------------------------
+  /// @brief      Register a callback to be invoked when the framework has
```
Registers

stuartmorgan commented on lib/ui/window/platform_configuration.h:
```diff
@@ -430,6 +464,11 @@ class PlatformConfiguration final {
   std::unordered_map<int, fml::RefPtr<PlatformMessageResponse>>
       pending_responses_;
+
+  // We use id 0 to mean that no response is expected.
```
s/We use id 0 to mean/An ID of 0 indicates/

stuartmorgan commented on lib/ui/window/window.h:
```diff
@@ -8,7 +8,9 @@
 #include <string>
 #include <unordered_map>
 #include <vector>
+#include <functional>
```
Nit: alphatize

stuartmorgan commented on lib/ui/window/window.h:
```diff
@@ -27,6 +29,7 @@ class Window final {
   const ViewportMetrics& viewport_metrics() const { return viewport_metrics_; }
 
   void DispatchPointerDataPacket(const PointerDataPacket& packet);
+  void DispatchKeyDataPacket(const KeyDataPacket& packet, uint64_t response_id);
```
This should have a comment (as should all the others, but they are out of scope for this PR).

stuartmorgan commented on shell/platform/embedder/embedder.cc:
```diff
@@ -1492,6 +1492,73 @@ FlutterEngineResult FlutterEngineSendPointerEvent(
+// Many platforms assert the character to be less than 2 int16's, i.e. 4 bytes,
+// therefore the character data is asserted to be less than double the amount,
+// i.e. 8 bytes.
+constexpr size_t kKeyEventCharacterMaxBytes = 8;
```
`static constexpr`

Or just declare it in `FlutterEngineSendKeyEvent`

stuartmorgan commented on shell/platform/embedder/embedder.cc:
```diff
@@ -1492,6 +1492,73 @@ FlutterEngineResult FlutterEngineSendPointerEvent(
+
+  auto response = [callback, user_data](bool handled) {
+        callback(handled, user_data);
+      };
```
Indentation of these lines is strange.

stuartmorgan commented on shell/platform/embedder/embedder.cc:
```diff
@@ -1492,6 +1492,73 @@ FlutterEngineResult FlutterEngineSendPointerEvent(
+  packet->SetKeyData(key_data);
+  packet->SetCharacter(character);
+
+  auto response = [callback, user_data](bool handled) {
```
I would expect a null callback to be allowed; some embedders (e.g., something running full screen on an embedded device) aren't going to care if it's been handled.

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -611,6 +611,65 @@ typedef struct {
+  kFlutterKeyEventKindRepeat,
+} FlutterKeyEventKind;
+
+/// A structure to represent a change of state of a key.
```
s/change of state of a key/key event/

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -611,6 +611,65 @@ typedef struct {
+  /// immediately before the actual event.
+  ///
+  /// An event being synthesized means that the framework will not trust the
+  /// `timestamp` of the event.
```
I'm not sure what this means to a client of this function. It suggests that maybe `timestamp` doesn't need to be set for synthesized events, but I don't think that's actually true. Maybe instead say something like "A synthesized event should use the timestamp of the time at which it was created, since the missed event's actual timestamp is unknown."

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -1549,6 +1608,33 @@ FlutterEngineResult FlutterEngineSendPointerEvent(
     size_t events_count);
 
+//------------------------------------------------------------------------------
+/// @brief      Send a key event to the engine, causing the framework to
```
Sends

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -1549,6 +1608,33 @@ FlutterEngineResult FlutterEngineSendPointerEvent(
 
+//------------------------------------------------------------------------------
+/// @brief      Send a key event to the engine, causing the framework to
+///             dispatch a key event that is simply transformed from this one.
```
I would change the whole sentence to just "Sends a key event to the engine." The fact that it's internally translated and dispatched is, I think, pretty obvious.

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -1549,6 +1608,33 @@ FlutterEngineResult FlutterEngineSendPointerEvent(
+///                            longer access `event` after returning.
+/// @param[in]  callback       The callback invoked by the engine when the
+///                            Flutter application has decided whether it handles
+///                            this event.
```
If you do allow null, document that here.

stuartmorgan commented on shell/platform/embedder/embedder.h:
```diff
@@ -1549,6 +1608,33 @@ FlutterEngineResult FlutterEngineSendPointerEvent(
+///                            this event.
+/// @param[in]  user_data      The context associated with the callback. The exact
+///                            same value will used to invoke `callback`. Accepts
+///                            nullptr or a non-pointer value.
```
Remove "or a non-pointer value"; it sounds like sending a pointer value is not allowed. Someone who knows what they are doing enough to pass data around as a `void *` is already familiar with that pattern and doesn't need to be told they can do it.

stuartmorgan commented on shell/platform/embedder/embedder_engine.h:
```diff
@@ -60,6 +60,9 @@ class EmbedderEngine {
   bool DispatchPointerDataPacket(
       std::unique_ptr<flutter::PointerDataPacket> packet);
 
+  bool DispatchKeyDataPacket(std::unique_ptr<flutter::KeyDataPacket> packet,
```
Comment.

stuartmorgan commented on shell/platform/embedder/tests/embedder_unittests.cc:
```diff
@@ -1191,5 +1191,205 @@ TEST_F(EmbedderTest, CanLaunchAndShutdownWithAValidElfSource) {
+  }
+}
+
+void expect_key_event_eq(const FlutterKeyEvent& subject, const FlutterKeyEvent& baseline) {
```
Comment.

stuartmorgan replied:
`ExpectKeyEventEq`

stuartmorgan commented on shell/platform/embedder/tests/embedder_unittests.cc:
```diff
@@ -1191,5 +1191,205 @@ TEST_F(EmbedderTest, CanLaunchAndShutdownWithAValidElfSource) {
+  bool returned;
+} KeyEventUserData;
+
+FlutterKeyEventKind unserializeKeyEventKind(uint64_t kindInt) {
```
Comment.

stuartmorgan replied:
s/unserialize/Unserialize/

stuartmorgan replied:
s/`kindInt`/`kind`/

stuartmorgan commented on shell/platform/embedder/tests/embedder_unittests.cc:
```diff
@@ -1191,5 +1191,205 @@ TEST_F(EmbedderTest, CanLaunchAndShutdownWithAValidElfSource) {
+  ready.Wait();
+
+  // A normal down event
+  const FlutterKeyEvent downEventUpperA {
```
`down_event_upper_a`

stuartmorgan commented on shell/platform/embedder/tests/embedder_unittests.cc:
```diff
@@ -1191,5 +1191,205 @@ TEST_F(EmbedderTest, CanLaunchAndShutdownWithAValidElfSource) {
+  EXPECT_EQ(echoed_char, 0x41llu);
+
+  // A repeat event with multi-byte character
+  const FlutterKeyEvent repeatEventWideChar {
```
`repeat_event_wide_char`

stuartmorgan commented on shell/platform/embedder/tests/embedder_unittests.cc:
```diff
@@ -1191,5 +1191,205 @@ TEST_F(EmbedderTest, CanLaunchAndShutdownWithAValidElfSource) {
+  EXPECT_EQ(echoed_char, 0x2206llu);
+
+  // An up event with no character, synthesized
+  const FlutterKeyEvent upEvent {
```
`up_event`

stuartmorgan reviewed (changes requested):
Mostly nits, but a few more substantive comments. Basically looks good though.

I only really reviewed the C++, not the Dart side, since I'm not familiar enough with that area of the code to give a good review.

dkwingsmt commented:
@mdebbar Can you take another look at the Web part? I should have addressed your issues. (I'm still fixing Stuart's comments but they're not related to web anyway).

mdebbar commented on lib/web_ui/lib/src/engine/keyboard_binding.dart:
```diff
@@ -0,0 +1,487 @@
+  bool handleEvent(FlutterHtmlKeyboardEvent event) {
+    final Duration timeStamp = _eventTimeStampToDuration(event.timeStamp!);
+
+    if (_shouldPreventDefault(event)) {
+      event.preventDefault();
```
Reminder to incorporate Greg's changes here.

dkwingsmt replied:
Addressed! Thanks for the reminder.

mdebbar commented on lib/web_ui/lib/src/engine/platform_dispatcher.dart:
```diff
@@ -170,6 +172,34 @@ class EnginePlatformDispatcher extends ui.PlatformDispatcher {
     invoke1<ui.PointerDataPacket>(_onPointerDataPacket, _onPointerDataPacketZone, dataPacket);
   }
 
+  /// A callback that is invoked when pointer data is available.
```
nit: "when key data"

mdebbar reviewed (approved):
Looks good to me other than the `preventDefault` comment.

dkwingsmt reviewed (commented):
Resolved all comments.

dkwingsmt commented:
@stuartmorgan Can you take another look?

mdebbar commented on lib/web_ui/lib/src/engine/keyboard_binding.dart:
```diff
@@ -0,0 +1,479 @@
+  bool _onKeyData(ui.KeyData data) {
+    bool? result;
+    EnginePlatformDispatcher.instance.invokeOnKeyData(data,
+      (bool handled) { result = handled; });
```
Is this callback called synchronously? I'm worried that `_onKeyData` may return before the callback gets a chance to update `result`.

dkwingsmt replied:
It is supposed to. If the code is returned before the handler is called, `result` will be null and the code will throw on the `!` of the return statement. Let me add a comment to explain it.

stuartmorgan reviewed (approved):
The changes all LGTM!
